File: Reqnroll/Chapter19/19.3-TestLogPattern/Before/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.App.Models;
using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(OrderingApiDriver orderingApiDriver)
{
    private TestActionResult<Order> placeOrderResult = TestActionResult<Order>.NotExecuted;

    [When("they place an order for {int} pizzas of size {PizzaSize}")]
    public async Task WhenTheyPlaceAnOrderForPizzasOfSize(int count, PizzaSize size)
    {
        var orderRequest = new PlaceOrderRequestObjectMother()
            .WithItems(count, size: size)
            .Build();
        placeOrderResult = await orderingApiDriver
            .PlaceOrder(orderRequest)
            .AttemptExecute();
    }

    [Then("the order should be rejected with message {string}")]
    public void ThenTheOrderShouldBeRejectedWithMessage(string expectedMessage)
    {
        placeOrderResult.AssertFailedWithErrorMessageContains(expectedMessage);
    }
}

File: CucumberJVM/Chapter19/19.4-EmbeddedApplicationLogPattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.Order;
import com.wimp.app.models.PizzaSize;
import com.wimp.app.specs.drivers.OrderingApiDriver;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import com.wimp.app.specs.support.TestActionResult;
import io.cucumber.java.en.*;

public class OrderingStepDefinitions {
    private final OrderingApiDriver orderingApiDriver;

    private TestActionResult<Order> placeOrderResult = TestActionResult.notExecuted();

    public OrderingStepDefinitions(OrderingApiDriver orderingApiDriver) {
        this.orderingApiDriver = orderingApiDriver;
    }

    @When("they place an order for {int} pizzas of size {pizza-size}")
    public void theyPlaceAnOrderForPizzasOfSize(int count, PizzaSize size) throws Exception {
        var orderRequest = new PlaceOrderRequestObjectMother()
            .withItems(count, size)
            .build();
        placeOrderResult = orderingApiDriver.placeOrder(orderRequest).attemptExecute();
    }

    @Then("the order should be rejected with message {string}")
    public void theOrderShouldBeRejectedWithMessage(String expectedMessage) {
        placeOrderResult.assertFailedWithErrorMessageContains(expectedMessage);
    }
}
